crates/re_ui/examples/re_ui_example.rs

```diff
@@ -600,3 +645,788 @@ impl egui_tiles::Behavior<Tab> for MyTileTreeBehavior {
+                let anything_being_decidedly_dragged = ui
+                    .memory(|mem| mem.is_anything_being_dragged())
+                    && ui.input(|i| i.pointer.is_decidedly_dragging());
+                if anything_being_decidedly_dragged {
+                    let (top, bottom) = response.rect.split_top_bottom_at_fraction(0.5);
+
+                    let (insert_y, target) = if ui.rect_contains_pointer(top) {
+                        (Some(top.top()), Some(i))
+                    } else if ui.rect_contains_pointer(bottom) {
+                        (Some(bottom.bottom()), Some(i + 1))
+                    } else {
+                        (None, None)
+                    };
+
+                    if let Some(insert_y) = insert_y {
+                        ui.painter().hline(
+                            ui.cursor().x_range(),
+                            insert_y,
+                            (2.0, egui::Color32::WHITE),
+                        );
+
+                        // TODO(emilk/egui#3841): it would be nice to have a drag specific API for that
+                        if ui.input(|i| i.pointer.any_released()) {
+                            target_item_pos = target;
+                        }
+                    }
+                }
+            }
+
+            //
+            // Handle the swap command (if any)
+            //
+
+            if let (Some(source), Some(target)) = (source_item_pos, target_item_pos) {
+                if source != target {
+                    let item = self.items.remove(source);
+
+                    if source < target {
+                        self.items.insert(target - 1, item);
+                    } else {
+                        self.items.insert(target, item);
+                    }
+                }
+            }
+        }
+    }
+}
+
+// ==============================================================================
+// HIERARCHICAL DRAG AND DROP DEMO
+
+mod hierarchical_drag_and_drop {
+    use std::collections::{HashMap, HashSet};
+
+    use egui::NumExt;
+
+    use re_ui::ReUi;
+
+    #[derive(Hash, Clone, Copy, PartialEq, Eq)]
+    struct ItemId(u32);
+
+    impl ItemId {
+        fn new() -> Self {
+            Self(rand::random())
+        }
+    }
+
+    impl std::fmt::Debug for ItemId {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "#{:04x}", self.0)
+        }
+    }
+
+    impl From<ItemId> for egui::Id {
+        fn from(id: ItemId) -> Self {
+            Self::new(id)
+        }
+    }
+
+    enum Item {
+        Container(Vec<ItemId>),
+        Leaf(String),
+    }
+
+    #[derive(Debug)]
+    enum Command {
+        /// Set the selection to the given item.
+        SetSelection(ItemId),
+
+        /// Toggle the selected state of the given item.
+        ToggleSelected(ItemId),
+
+        /// Move the currently dragged item to the given container and position.
+        MoveDraggedItemTo(ItemId, usize),
```

Comment:
How does the receiver know what is "currently dragged"?

```suggestion
        MoveItem {
            what_to_move: ItemId,
            target_container: ItemId,
            target_row: usize,
        }
```